### ISSUE TITLE
[MNG-5082] Regression: Could not resolve dependencies for project

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7128]
**[Florian Brunner](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=puce)** opened **[MNG-5082](https://issues.apache.org/jira/browse/MNG-5082?redirect=false)** and commented

Sometimes, when you need a jar, which is not available from a repository, and for some reason you don't want to deploy it to the corporate repository (yet), you can do something like this:

```xml
    <build>
     <plugins>
                   ...
          <plugin>
               <artifactId>maven-install-plugin</artifactId>
               <executions>
                        <execution>
                                 <id>some-lib</id>
                                 <phase>validate</phase>
                                 <goals>
                                          <goal>install-file</goal>
                                 </goals>
                                 <configuration>
                                          <file>lib/some-lib-1.1.jar</file>
                                          <groupId>org.someorg.somelib</groupId>
                                          <artifactId>some-lib</artifactId>
                                          <version>1.1</version>
                                          <packaging>jar</packaging>
                                 </configuration>
                        </execution>
               </executions>
          </plugin>
          ...
     </plugins>
</build>
```

And then you can declare a dependency on this lib like this:

```xml
    <dependencies>
           ...
     <dependency>
          <groupId>org.someorg.somelib</groupId>
          <artifactId>some-lib</artifactId>
          <version>1.1</version>
     </dependency>
     ...
</dependencies>
```

This worked fine with `Maven 2.2.1` but not with `Maven 3.0.3` -> regression!

```
Error message:
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.951s
[INFO] Finished at: Thu Apr 28 15:53:15 BST 2011
[INFO] Final Memory: 4M/121M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project myproject: Could not resolve dependencies for project someGroupId:someArtifactId:jar:0.1-SNAPSHOT: Could not find artifact org.someorg.somelib:some-lib:jar:1.1 in nexus (<nexus url>) -> [Help 1]
[ERROR] 
```

There is a work-around, which at least works in some cases:

```
mvn clean validate
mvn install
```

Some notes on this work-around:
- install should call validate (which it does, but apparently with some side-effects). It should not be necessary to call validate explicitly.
- "mvn clean validate install" does not work! It shouldn't matter if you call "mvn clean validate install" or "mvn clean validate" and "mvn install". Both should have the same result.
- I currently don't see a way to specify in Jenkins/ Hudson to first execute "clean validate" and then "install" -> broken build on continuous integration server which has to be fixed manually
- This work-around breaks the Maven way, because you cannot simply go to a project anymore and just call "mvn clean install" to build it.

---

**Affects:** 3.0.3

**Attachments:**
- [MNG-5082-maven-core.patch](https://issues.apache.org/jira/secure/attachment/12713116/MNG-5082-maven-core.patch) (_3.04 kB_)

3 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Robin Howlett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rhowlett)** commented

I concur this is a regression blocker.

I have the same scenario and have found that I must run the specified phase that the install-file goal is attached on its own, for example:

> mvn clean; mvn package

will work; the "clean" phases will install the file (as the execution is tied to the clean goal) - I can then run "package".

However, this will fail with Maven 3:

> mvn clean package

When running under Maven 2.2, no such error is encountered and the file is installed to my local repo.

**Comment by jira-importer** [comment]
**[Florian Brunner](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=puce)** commented

I stumbled again over the same issue. Any updates?

**Comment by jira-importer** [comment]
**[Robert Balazsi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rbalazsi)** commented

I think I've managed to fix the problem, I attached a patch.

The method "MojoExecutor.execute(MavenSession session, List\<MojoExecution> mojoExecutions, ProjectIndex projectIndex)" is called for every plugin to be executed. However, the method collected the dependencies for the entire lifecycle, not just the current phase. So before the "compile" phase, the library dependency ("some-lib" in the issue description) would be satisfied by the "maven-install-plugin" in the "validate" phase. However, it is attempted to be resolved beforehand (at the beginning of the lifecycle).

I solved it by rewriting the execute method so that the "newDependencyContext()" method would be called for every MojoExecution object, instead of only once for the list of MojoExecution objects.
I'm not sure how to write unit/integration tests, though, as the scenario uses the "maven-install-plugin", I appreciate any suggestions. All of the current tests pass, of course.

This is my first contribution, I hope that it turns out to be useful. :-)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

IMHO all dependencies which are not part of the reactor should be available at startup of the build, which is required to make a valid buildPlan. So even though this used to work with Maven2, I think the approach of Maven3 is better, since Maven can't detect which plugins will effect the buildPlan at startup. I'd like to close this as `Won't Fix`.
Doing an `install-file` during `validate` means that it'll be execute during every build, installing the same file over and over again. That's also a sign that this approach can't be valid.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Based on comments of Robert Scholte this will be closed.
